junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/discovery/ClassOrderingVisitor.java

```diff
@@ -0,0 +1,104 @@
+
+/**
+ * @since 5.8
+ */
+class ClassOrderingVisitor implements TestDescriptor.Visitor {
+
+	private static final Logger logger = LoggerFactory.getLogger(ClassOrderingVisitor.class);
+
+	private final JupiterConfiguration configuration;
+
+	public ClassOrderingVisitor(JupiterConfiguration configuration) {
+		this.configuration = configuration;
+	}
+
+	@Override
+	public void visit(TestDescriptor testDescriptor) {
+		Optional<ClassOrderer> testClassOrderer = configuration.getTestClassOrderer();
+		if (!testClassOrderer.isPresent()) {
+			return;
+		}
+		if (testDescriptor instanceof JupiterEngineDescriptor) {
+			JupiterEngineDescriptor jupiterEngineDescriptor = (JupiterEngineDescriptor) testDescriptor;
+			try {
+				orderContainedClasses(jupiterEngineDescriptor, testClassOrderer.get());
+			}
+			catch (Throwable t) {
+				UnrecoverableExceptions.rethrowIfUnrecoverable(t);
+				logger.error(t, () -> "Failed to order classes");
+			}
+		}
+	}
+
+	private void orderContainedClasses(JupiterEngineDescriptor jupiterEngineDescriptor, ClassOrderer classOrderer) {
+		Set<? extends TestDescriptor> children = jupiterEngineDescriptor.getChildren();
+
+		List<TestDescriptor> nonClassTestDescriptors = children.stream()//
+				.filter(testDescriptor -> !(testDescriptor instanceof ClassBasedTestDescriptor))//
+				.collect(Collectors.toList());
+
+		List<DefaultClassDescriptor> classDescriptors = children.stream()//
+				.filter(ClassBasedTestDescriptor.class::isInstance)//
+				.map(ClassBasedTestDescriptor.class::cast)//
+				.map(DefaultClassDescriptor::new)//
+				.collect(toCollection(ArrayList::new));
+
+		// Make a local copy for later validation
+		Set<DefaultClassDescriptor> originalClassDescriptors = new LinkedHashSet<>(classDescriptors);
+
+		classOrderer.orderClasses(new DefaultClassOrdererContext(classDescriptors, this.configuration));
+
+		int difference = classDescriptors.size() - originalClassDescriptors.size();
+
+		if (difference > 0) {
+			logger.warn(() -> String.format("ClassOrderer [%s] added %s ClassDescriptor(s) which will be ignored.",
+				classOrderer.getClass().getName(), difference));
+		}
+		else if (difference < 0) {
+			logger.warn(() -> String.format(
+				"ClassOrderer [%s] removed %s ClassDescriptor(s) which will be retained with arbitrary ordering.",
+				classOrderer.getClass().getName(), -difference));
+		}
+
+		Set<TestDescriptor> sortedClassTestDescriptors = classDescriptors.stream()//
+				.filter(originalClassDescriptors::contains)//
+				.map(DefaultClassDescriptor::getTestDescriptor)//
+				.collect(toCollection(LinkedHashSet::new));
+
+		// Currently no way to removeAll or addAll children at once.
+		Stream.concat(sortedClassTestDescriptors.stream(), nonClassTestDescriptors.stream())//
+				.forEach(jupiterEngineDescriptor::removeChild);
+		Stream.concat(sortedClassTestDescriptors.stream(), nonClassTestDescriptors.stream())//
+				.forEach(jupiterEngineDescriptor::addChild);
```

Comment:
There's quite some duplication with `MethodOrderingVisitor`. Have you already tried avoiding that?

Reply:
I have pushed a distinct commit to discuss around reducing duplication between these two classes.

But as often in this cases, I am not sure the resulting piece is clearer, wdyt ?

Reply:
I see what you mean. I still think it's better to avoid the duplication in this case.